checker/src/call_graph.rs

```diff
@@ -0,0 +1,770 @@
+use petgraph::visit::Bfs;
+use petgraph::{Direction, Graph};
+use rustc_hir::def_id::DefId;
+use serde::{Deserialize, Serialize};
+use serde_json;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::fs;
+use std::path::Path;
+
+// An unique identifier for a Rust type string.
+type RTypeIdx = u32;
+
+// A unique identifier for a graph node.
+type NodeIdx = NodeIndex<DefaultIx>;
+
+// Convenience types for the graph folding algorithm.
+type HalfRawEdge = (NodeIdx, RTypeIdx);
+type RawEdge = (NodeIdx, NodeIdx, RTypeIdx);
+type MidpointExcludedMap = HashMap<NodeIdx, (HashSet<HalfRawEdge>, HashSet<HalfRawEdge>)>;
+
+/// Specifies reduction operations that may be perfomed
+/// on a call graph. Supported operations are:
+#[derive(Debug, Clone, Serialize, Deserialize)]
+enum CallGraphReduction {
+    /// Only include nodes reachable from the given node.
+    /// See `CallGraph::filter_reachable`.
+    Slice(String),
+    /// Remove nodes in the graph that belong to crates other than
+    /// `CallGraphConfig.included_crates`. The outgoing edges of these
+    /// removed node are connected to the node's parents.
+    /// See `CallGraph::fold_excluded`.
+    Fold,
+    /// Remove duplicated edges (only considers edge endpoints).
+    /// See `CallGraph::deduplicate_edges`.
+    Deduplicate,
+    /// Remove nodes that have no incoming or outgoing edges.
+    /// See `CallGraph::filter_no_edges`.
+    Clean,
+}
+
+/// Configuration options for call graph generation.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct CallGraphConfig {
+    /// Optionally specifies location for graph to be output in dot format
+    /// (for graphviz).
+    dot_output_path: Option<String>,
+    /// Optionally specifies location for graph to be output as datalog input
+    /// relations.
+    ddlog_output_path: Option<String>,
+    /// Optionally specifies location for mapping from type identifiers to
+    /// type strings.
+    type_map_output_path: Option<String>,
+    /// Optionally specifies the location for manually defined type relations
+    /// to be imported.
+    type_relations_path: Option<String>,
+    /// A list of call graph reductions to apply sequentially
+    /// to the call graph.
+    reductions: Vec<CallGraphReduction>,
+    /// A list of crates to include in the call graph.
+    /// Nodes belonging to crates not in this list will be removed.
+    included_crates: Vec<String>,
+}
+
+impl Default for CallGraphConfig {
+    fn default() -> CallGraphConfig {
+        CallGraphConfig {
+            dot_output_path: None,
+            ddlog_output_path: None,
+            type_map_output_path: None,
+            type_relations_path: None,
+            reductions: Vec::<CallGraphReduction>::new(),
+            included_crates: Vec::<String>::new(),
+        }
+    }
+}
+
+/// The type of a call graph node.
+#[derive(Debug, Clone, PartialEq)]
+enum NodeType {
+    /// Regular root
```

Comment:
It is not clear to me what the difference is between a regular root and a crate root. Can you amplify the comments to make clear the point of this design?

Reply:
A crate root is a public function of a crate as visited by the `crate_visitor`. The regular root might be a poor name; anything that isn't a crate root is considered to be a regular root. I'm opening to changing it to something clearer, and I will clarify the comments.